keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,859 @@
+3. The keys and values within the `node.status.capabilities` map must adhere to the same validation rules as standard Kubernetes labels. The Kubelet validates each capability and discards any key-value pair that is invalid, logging an error.
+
+**Note:**  
+
+*   We currently have [Node Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6279) and [Node Runtime Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6251) which publish some runtime features through Node Status. They are too narrowly scoped and currently not used for scheduling pods. We can deprecate those fields and introduce them into NodeCapabilities if required. 
+*   Other API options are explored in the [Alternatives](#alternatives) section.
+
+
+### kube-scheduler Changes
+
+To enable capability-based scheduling, this KEP proposes a new scheduler plugin named NodeCapabilityFilter. This plugin will be part of the default scheduling framework and will be enabled when the NodeCapabilities feature gate is active. DaemonSet Pods will be skipped by this plugin.
+
+#### Plugin Implementation
+
+`NodeCapabilityFilter` plugin would implement two extension points in the [scheduling framework](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/)
+
+1. Prefilter: In this stage, the plugin will inspect the PodSpec to infer the set of required NodeCapability key-value pairs. This computed set of requirements is then stored in the prefilter state.
+2. Filter: For each node being evaluated, this phase retrieves the pre-computed requirements and checks them against the node's `node.status.capabilities` map. If any required capability is not present or the value does not match on the node, the plugin will filter out the node, making it unschedulable for the pod. If no nodes satisfy the capability requirements, the pod will remain in a pending state with a FailedScheduling event detailing the reason.
+
+#### Performance Considerations
+
+Introducing this plugin adds a new step to the scheduling cycle, which may have an impact on scheduling throughput. However, this trade-off is acceptable because it prevents the greater inefficiency of scheduling a pod onto a node where it cannot run.
+
+
+### Admission Controller Changes
+
+To enable the validation of API requests against Node Capabilities, this KEP proposes the introduction of a new admission controller plugin `NodeCapabilityValidator `that will be enabled when the NodeCapabilities feature gate is active. This is necessary because the validation logic that requires cross-referencing objects i,e looking up the `Node` object that a `Pod` is bound to (spec.nodeName). The admission controller would 
+
+*   Inspects incoming pod update request and infer if it depends on any node capability. 
```

Comment:
I'd love if if the admission controller could share code with the scheduler here so infererencing is happening consistently. I'd really love it if they could just see in the API directly that thefeature is enabled with a feature gate, and match against that feature gate, but I don't know if the apiserver encodes that information once it's serialized

Reply:
> There are various components that can mess up the capabilities.

Do you mean by updating the node status? Or do you mean changing node configuration?

> So we need to declare the behavior. If kubelet is the ultimate source of capabilities - it must continuously reconcile it.

Again, do you mean reconcile the node status, or continuously recompute the set of capabilities?

> And must not believe capabilities API server has if kubelet doesn't have them defined.

I don't understand this part. Do you mean the Kubelet needs its own internal source of truth for capabilities, rather than just reading them back from the node object?